File: Chanseok/프로그래머스/Lv4/17685. ［3차］ 자동완성/자동완성.java
import java.util.Arrays;

class Solution {
    public int solution(String[] words) {
        int answer = 0;

        Arrays.sort(words);

        int wordsLen = words.length;
        String prev, cur, next;
        int prevCnt, nextCnt;
        for(int i = 0; i < wordsLen; ++i) {

            cur = words[i];

            prevCnt = 0;
            if(i != 0) {
                prev = words[i-1];
                for(int j = 0; j < prev.length(); ++j) {
                    if(cur.charAt(j) != prev.charAt(j)) {
                        prevCnt = j+1;
                        break;
                    }
                }
                if(prevCnt == 0) prevCnt = prev.length()+1;
            }
            nextCnt = 0;
            if(i != wordsLen-1) {
                next = words[i+1];
                for(int j = 0; j < cur.length(); ++j) {
                    if(cur.charAt(j) != next.charAt(j)) {
                        nextCnt = j+1;
                        break;
                    }
                }
                if(nextCnt == 0) nextCnt = cur.length();
            }

            answer += Math.max(prevCnt, nextCnt);
        }

        return answer;
    }
}

File: Seonguk/프로그래머스/자동완성/자동완성.py
def make_trie(trie, words):
    for word in words:
        cur=trie
        l=len(word)
        for w in word:
            if w in cur:
                cur=cur[w]
                cur['!'].append(l)
            else:
                cur[w]={}
                cur = cur[w]
                cur['!']=[l]

def search_trie(trie, word):
    if len(trie[word[0]]['!']) == 1:
        return 1

    cnt=0
    for w in word:
        if w in trie:
            trie=trie[w]
            cnt+=1
        if len(trie['!']) == 1:
            return cnt
        
    return len(word)
        
def solution(words):
    answer = 0
    trie={}
    make_trie(trie, words)
    for word in words:
        answer+=search_trie(trie, word)
    
    return answer
